traffic_ops/app/db/migrations/2022021523082100_server_config_update.up.sql

```diff
@@ -0,0 +1,28 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership.  The ASF
+ * licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+CREATE TABLE IF NOT EXISTS public.server_config_update ( 
```

Comment:
There's no need for one-to-many relationships - in fact that should be prohibited - so why not make these just columns on the existing `public.server` table?

Reply:
This is actually a 1:1 relationship. The `server_id` is a Primary Key, guaranteeing unique-ness, among other things. Then with the foreign key constraint it prevents the 1:n relationship. This requires _upserts_ rather than updates.

There are pros and cons to making this a separate table. A pro, for example, is the separation of indexes (or indicies). If this is a highly volatile table (or columns if it were in the `server` table) that required indexing, this could cause performance issues during writes.

That said, I was following a design decision made previously.

Reply:
Well, because it's a separate table it's actually a 1:0-or-1 relationship. There's no enforcement that a server must have a corresponding row in this table, and it's not possible to make that enforcement in the database, since you can't create a row with a foreign key to a row that doesn't exist. Those restrictions would have to be enforced by API code, which can be modified in arbitrary ways by plugins, and bad data could be introduced in future migrations since those would bypass anything we build in TO. For those reasons I think columns on the existing table are more maintainable than a separate table.

> I was following a design decision made previously

Are you referring to previous changes made to servers, or some design discussion somewhere regarding this feature in particular?

Reply:
You're right, it is a 1:0-or-1 table. And that's a good point about the lack of enforcement, but I believe that is ok. Out of curiosity, are there servers or server configurations that may not (or ever) need to have a config updated or a reval updated? For example, would servers designated as TRs, TO DBs, Influx, RIAK/TV, etc all require `updPending` or `revalPending`? That would be a potential and objective reason to have a 1:0-or-1 relationship.

Reply:
That _could_ be a reason to do that. There are other changes we could make in those cases that would also greatly simplify them, too. Like Traffic Monitors don't need to have multiple network interfaces. We don't even need to catalogue which network interface it uses at all. No part of ATC relies on that information. The trouble is that TMs are servers, and servers have interfaces. We _could_ make the typing really complex, but that gets very difficult especially when you consider that "type" is a mutable data member in ATC. For example, we can't say:
```typescript
// "BaseServer" contains all the other fields we're not interested in
type Server = BaseServer & ({
	type: "EDGE" | "MID",
	interfaces: Array<{
		name: string;
		monitor: boolean;
		mtu: number | null;
		addresses: Array<{
			address: string;
			subnet: string | null;
			isServiceAddress: boolean;
		}>;
	}>;
} | {
	type: "TRAFFIC_MONITOR";
	ipAddresses: Array<string>;
});
```
... because "type" isn't actually a property of a server. In fact, in production deployments the vast majority of cache servers do *not* use the "EDGE" or "MID" type, and I'm sure many Traffic Monitors are actually using a type named "RASCAL" or something. The IDs of those types can't be depended on and the names can't be depended on, so the only thing we could do is handle the cases where each type has either way of specifying IP addresses.

You can see how that quickly becomes a nightmare, both to maintain and to document/explain - to say nothing of how annoying it would be to try to make Go structures that model such a widely varying type.

In essence what I'm trying to say is the only really good way to do that, IMO, would be to separate things that are cache servers out from things that are just "servers", so that they use different API endpoints to define their interactions - and at that point you're almost certainly going to have them in separate database tables, otherwise it's still a nightmare to work with for "Type" reasons.

Reply:
> Well, because it's a separate table it's actually a 1:0-or-1 relationship. There's no enforcement that a server must have a corresponding row in this table,

That's okay by design. By using a foreign key, it's easy to ensure rows on the new table require a corresponding row in the server table. And the feature is intentionally designed so that nulls are ok, and coalesce to a zero time by the application, and work appropriately.

By putting it on the server table, it requires all old rows be migrated. Which can easily be made to work, the migration can simply also be the zero time. 

But it's perfectly acceptable for a Server to not have an update time, and legitimate from a data perspective, meaning "this server has never been queued" and "this server has never applied configuration" respectively. Putting it on servers is less legitimate from a data perspective. We're now creating artificial sentinel values, for data that is nonexistent. It's more correct for the schema to reflect the nonexistence. In technical terms, the separate table is more Normal.

It's also less Normal server types that can't be queued, i.e. non-caches.

That said, I don't feel that strongly about it, and I see Taylor already changed the code. Meh.

Reply:
> That's okay by design. By using a foreign key, it's easy to ensure rows on the new table require a corresponding row in the server table. And the feature is intentionally designed so that nulls are ok, and coalesce to a zero time by the application, and work appropriately.

> By putting it on the server table, it requires all old rows be migrated. Which can easily be made to work, the migration can simply also be the zero time.

All of the coalescence is being done in queries anyway, so you wouldn't need to migrate existing rows either way.`NULL` is still `NULL`.

> We're now creating artificial sentinel values, for data that is nonexistent. It's more correct for the schema to reflect the nonexistence. 

I absolutely agree, but I don't see why a `NULL` value can't indicate the lack of an existing value; that's it's purpose.